Automattic/wp-calypso
# Add touch device handling for masterbar items with submenu items.

Related to # - This PR is aimed to merge into https://github.com/Automattic/wp-calypso/pull/92472/, as it is a subgoal for that PR.

## Proposed Changes

Adds touch handling to the menu items:
* Adds a `isOpenByTouch` state property. When present an `is-open` class is added to the component.
* Adds the `.is-open` selector on the css for the hover state that shows the submenu items.
* Add a touch event onto the parent menu item that prevents the default behavior and toggles the menu open state/selector to see the items.
* Adds touch events on the submenu items to ensure the menu is then closed and proper navigation/click handling still fires.
* Adds a listener to the document for `touchstart` events. If a `touchstart` happens outside of the element, we ensure the menu closes.
    * Adds a ref (and sometimes a wrapper div noted in comments) so this callback can determine if the touch event started from within the component.

BEFORE (touch flow)
![touch-before](https://github.com/Automattic/wp-calypso/assets/28742426/fbb0b1a2-8d57-416d-b015-1bda582c0327)

AFTER (touch flow)
![touch-flow-after](https://github.com/Automattic/wp-calypso/assets/28742426/de0c0231-3d31-47cc-8de3-6c6eac2feff1)



## Why are these changes being made?


* Need to support touch device flows with the new masterbar.

## Testing Instructions



* Test the menu items in mouse flows and verify they continue to work as expected.
* Open devtools and simulate a touch device.
* Verify that clicking on a menu item with sub items toggles the menu, instead of just firing that anchor/button.
* Verify clicking a submenu item works as expected (navigates) and closes the menu.
* With a submenu open, verify clicking outside the component or opening another menu with submenu items closes the first open submenu.
* ... ?

## Pre-merge Checklist
